### PR TITLE
feat: Add `EvolComplexity` and `EvolQuality`

### LINKED ISSUE #296: [FEATURE] improve `tests` `text-descriptives` and `sentence-transformers`
**Is your feature request related to a problem? Please describe.**
These tests are slow and can better be mocked.

**Describe the solution you'd like**
Mock the returned values.

**Describe alternatives you've considered**
N.A.

**Additional context**
N.A.

### PR DESCRIPTION
**Pull request by davidberenstein1957**
Added `EvolComplexityTask` and `EvolQualityTask`.
- `EvolComplexityTask` is focused on improving the complexity of instructions and is a special implementation of `EvolInstruct` where we don't allow for the `breadth` evolution method.
- `EvolQualityTask` is focused on improving the complexity of `generation`/responses and is a more altered implementation of `EvolInstruct` with evolution methods for `["helpfulness", "relevance", "depth", "creativity", "details"]`.

Closes #296 
Closes #299

### CODE REVIEW

**Comment by davidberenstein1957** [comment]
docs are still WIP @plaguss

**Review by plaguss — COMMENTED** [review]
Nice! Just a reminder in the docstrings for one of the references. The only thing I'm not sure is if it makes sense to keep both `EvolInstruct` and `EvolComplexity`, to avoid complexity for the users.

File: src/distilabel/tasks/text_generation/evol_quality.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    This evolutionary process can be repeated for several rounds to obtain instruction data containing various complexities.
+    Currently the task is implemented as a single step, so to generate multiple evolutions you can "repeat" the instructions
+    in the original dataset. An example can be seen at the following script:
+    [examples/pipeline-evol-instruct-alpaca.py](https://github.com/argilla-io/distilabel/tree/main/examples/pipeline-evol-instruct-alpaca.py)
+
+    Args:
+        system_prompt (str, optional): the system prompt to be used. Not defined for this task.
+
+    References:
+        - [`WizardLM: Empowering Large Language Models to Follow Complex Instructions`](https://arxiv.org/abs/2304.12244)
```
Comment (plaguss) [inline src/distilabel/tasks/text_generation/evol_quality.py:71]: Update the reference to the Deita Paper

File: src/distilabel/tasks/text_generation/evol_complexity.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+from dataclasses import dataclass
+from typing import Any, Literal, Optional
+
+from distilabel.logger import get_logger
+from distilabel.tasks.base import get_template
+from distilabel.tasks.prompt import Prompt
+from distilabel.tasks.text_generation.evol_instruct import EvolInstructTask
+
+logger = get_logger()
```
Comment (gabrielmbmb) [inline src/distilabel/tasks/text_generation/evol_complexity.py:23]: we can remove the `logger` if it's not being used

File: src/distilabel/tasks/text_generation/evol_quality.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        return ["generations"]
+
+    def parse_output(self, output: str) -> Dict[str, List[str]]:
+        """Parses the output of the model into the desired format, applying the elimination step for bad generations.
+
+        Args:
+            output (str): the output of the model.
+
+        Note:
+            The eliminatin step is applied to the output, but only steps 2-4 in the paper are implemented.
```
Comment (gabrielmbmb) [inline src/distilabel/tasks/text_generation/evol_quality.py:131]:
```suggestion
            The elimination step is applied to the output, but only steps 2-4 in the paper are implemented.
```

File: src/distilabel/tasks/text_generation/evol_quality.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import sys
+
+if sys.version_info < (3, 9):
+    pass
+else:
+    pass
```
Comment (gabrielmbmb) [inline src/distilabel/tasks/text_generation/evol_quality.py:20]: remove this

File: src/distilabel/tasks/text_generation/evol_quality.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+from dataclasses import dataclass
+from typing import Any, Dict, List, Literal, Optional
+
+from distilabel.logger import get_logger
+from distilabel.tasks.base import get_template
+from distilabel.tasks.prompt import Prompt
+from distilabel.tasks.text_generation.evol_instruct import EvolInstructTask
+
+logger = get_logger()
```
Comment (gabrielmbmb) [inline src/distilabel/tasks/text_generation/evol_quality.py:30]: not used

File: src/distilabel/tasks/text_generation/evol_complexity.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+# inherits from evol-instruct with limited evolution methods
+_EVOL_COMPLEXITY_TEMPLATE = get_template("evol-instruct.jinja2")
+
+
+EvolutionMethod = Literal["constraints", "deepen", "concretizing", "reasoning"]
+
+
+@dataclass
+class EvolComplexityGeneratorTask(EvolInstructTask):
```
Comment (gabrielmbmb) [inline src/distilabel/tasks/text_generation/evol_complexity.py:34]: I would change the name of this class to just simply `EvolComplexityTask` or if we want to keep the reference of `generation` then I would align the name with the `TextGenerationTask` and call it `EvolComplexityGenerationTask`.

File: src/distilabel/tasks/text_generation/evol_quality.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    "helpfulness",
+    "relevance",
+    "depth",
+    "creativity",
+    "details",
+]
+
+
+@dataclass
+class EvolQualityGeneratorTask(EvolInstructTask):
```
Comment (gabrielmbmb) [inline src/distilabel/tasks/text_generation/evol_quality.py:46]: Same, I would call it `EvolQualityTask`

**Review by gabrielmbmb — CHANGES_REQUESTED** [review]
Hi! Looks good to me, I left some comments before merging

**Comment by gabrielmbmb** [comment]
Also, could we add a unit test covering the Evol Quality Task use of the method `_elimination_evolving` with `response_words`?

File: tests/tasks/text_generation/test_evol_tasks.py [hunk 8]
```diff
@@ -120,6 +116,16 @@ def test_evol_task(evolution_method: str, expected: str, instruct_type: object):
     assert isinstance(task, instruct_type)
     assert task.system_prompt == ""
     if isinstance(expected, str):
+        print("---COSA")
```
Comment (davidberenstein1957) [inline tests/tasks/text_generation/test_evol_tasks.py:119]: what is happening here?

Comment (plaguss) [reply tests/tasks/text_generation/test_evol_tasks.py:119]: ups, forgot to remove those

**Comment by plaguss** [comment]
> Also, could we add a unit test covering the Evol Quality Task use of the method `_elimination_evolving` with `response_words`?

Here it is: https://github.com/argilla-io/distilabel/pull/299/commits/7c309aa01dd79597adb597c52ee076c5ef19aa50. If the tests pass, and the docs are correct, I'll merge it

**Pull request merged**